robot_localization/pf.py

```diff
@@ -215,6 +246,24 @@ def update_particles_with_odom(self):
 
         # TODO: modify particles using delta
 
+        # create transformation matrix from odom to old neato reference frame
+        old_x_odom = old_odom_xy_theta[0]
+        old_y_odom = old_odom_xy_theta[1]
+        old_theta_odom = old_odom_xy_theta[2]
+        T_old_to_odom = np.array([[cos(old_theta_odom), -sin(old_theta_odom), old_x_odom], [
```

Comment:
So this seems like a transformation matrix and rotation matrix at the same time to translate between the last point-updating odometry frame and the current frame. 

A couple lines down to get the delta position we apply this rotation/translation to the current position of the odometry frame to get delta_position.

To avoid this transformation also applying the displacement of current_position to each point as well as the absolute rotation, won't we have to then subtract or somehow remove current_position from this output so that the transformation is from the coordinate frame origin? I guess I just don't understand what's going on in line 261 and why we can't just apply T_odom_to_old which seems to be the translation/rotation matrix to the individual particles and do the theta transformation that is already happening in the particle methods. It seems like we have a nonzero baseline from the incorporation of current_position but I think we could go through this in person.


All in all I think that a couple more comments like your other sections of code could help the clarity of this section. I may also be misunderstanding the math, I am rly tired with the conference stuff.
wouldn't this just get the final position of the current read odometry frame if the last not-recalculated movement was applied to the current frame (which, in reality has already experienced this transformation?). My intuition is that when applying T_odom_to_old we are applying the

Reply:
I hope that the comments cleared up what the transformation matrix is doing; I also think we should discuss this approach together next class. If I am understanding your question correctly, the transformation matrix is not actually moving any points, but returns the points' coordinates as they would be expressed in a different reference frame. Applying T_odom_to_old to the current_position (which is expressed in odom) will tell us where it is with respect to its old position. The good thing about this is that because this delta is from the neato's perspective, it can also be used as if it were from each particle's perspective. The final transformation is handled in the particles' update_pose method.